src/main/java/net/starlark/java/eval/StarlarkInt.java

```diff
@@ -436,17 +436,32 @@ public int compareTo(StarlarkInt x) {
 
   /** Returns a value whose signum is equal to x - y. */
   public static int compare(StarlarkInt x, StarlarkInt y) {
-    if (x instanceof Int32 && y instanceof Int32) {
-      return Integer.compare(((Int32) x).v, ((Int32) y).v);
+    long xl = 0, yl = 0;
+    boolean xbig = false, ybig = false;
+    try {
+      xl = x.toLongFast();
+    } catch (Overflow unused) {
+      xbig = true;
     }
-
     try {
-      return Long.compare(x.toLongFast(), y.toLongFast());
+      yl = y.toLongFast();
     } catch (Overflow unused) {
-      /* fall through */
+      ybig = true;
+    }
+
+    // If both arguments are big integers, we invoke BigInteger.compare.
+    // If neither argument is big integer, we compare longs.
+    // If only one argument is big integer, it is bigger than other if positive
+    // and smaller otherwise.
+    if (xbig && ybig) {
```

Comment:
FWIW, this version requires 4 branches to reach the common case (Long.compare), whereas the version I proposed used a control tree of depth two. Probably not a big deal, especially given the mystery and unpredictability of measurements using the JVM, but something to bear in mind when optimizing (especially in a language like C++ or Go or assembly).